[Wang Guanlin] iP

WangGLJoseph opened the pull request:
# Duke v0.9.0
> *"Be not afraid of growing slowly, be afraid of standing still."* - Chinese proverb

Duke offers you a:
- Hassle free
- Text based
- Minimalist
and completely **offline** experience of planning your tasks!

## Duke's features
- [x] Adding and saving tasks
- [x] GUI (coming soon!)
- [ ] Mystery features 😜

## Setting up Duke
1. Download it [here](https://github.com/WangGLJoseph/ip/releases/download/A-Jar/ip.jar)
2. Double-click to run it
3. Add your tasks
4. Enjoy **mastery** over your tasks 😆

A snippet of the workings of controlling your tasks. Behold, you've got the power!
   ```java
    public String finishTask(int index) throws DukeException {
        if (index > taskArrayList.size()) {
            throw new DukeException("This task index is not in the task list!");
        }
        taskArrayList.get(index - 1).markAsDone();
        return (sandwich("Congratulations! You have finished this task: "
                + taskArrayList.get(index - 1).toString()));
    }
   ```

Commit:
Add Gradle support

WangGLJoseph committed:
Level-1: Added Greet, Echo, Exit functions for Duke

WangGLJoseph committed:
Included gitignore file, ignore all files with .class extension

WangGLJoseph committed:
Level-2: Added Add, List functions for Duke

WangGLJoseph committed:
Created a Task class to represent tasks added by user

WangGLJoseph committed:
Changed Duke.java to represent userInputs as an ArrayList of Task objects instead of an ArrayList of String(s)

WangGLJoseph committed:
Removed deprecated code representing userInputs as an ArrayList of String(s)

WangGLJoseph committed:
Updated scanner to read userInput by using next() instead of nextLine()

WangGLJoseph committed:
Level-3: Added Mark as Done function for Duke

WangGLJoseph committed:
Added field for task type and getter method for task type icon

WangGLJoseph committed:
Updated instantiation of Task objects to include a taskType parameter

WangGLJoseph committed:
Abstracted out logic for beautifying task and list

WangGLJoseph committed:
Updated scanner to use nextLine after first command has been checked

WangGLJoseph committed:
Added constructor for Task that took in an additional string as reminder field

WangGLJoseph committed:
Updated todo and deadline cases to be recognizable as userInput

WangGLJoseph committed:
Added task reminder for Deadline tasks

WangGLJoseph committed:
Removed unnecesary use of delimiter for Deadline tasks

WangGLJoseph committed:
Updated listBeautify to use taskBeautify method

WangGLJoseph committed:
Level-4: Added ToDos, Events, Deadlines functions for Duke

WangGLJoseph committed:
Updated to use Inheritance to support multiple task types (ToDo, Event, Deadline)

WangGLJoseph committed:
Fixed bug where type of Event was displayed as [D]

WangGLJoseph committed:
A-TextUiTesting: Used the input/output redirection technique to semi-automate the testing of Duke

WangGLJoseph committed:
Removed deprecated method TaskBeautify() from Duke

WangGLJoseph committed:
Added DukeException to handle errors in Duke regarding invalid user inputs

WangGLJoseph committed:
Level-5: Added exceptions for unknown and incomplete user inputs

WangGLJoseph committed:
Level-6: Added Delete function to Duke, tasks stored in ArrayList<Task>

WangGLJoseph committed:
Updated text-ui-test to work with Duke Level-6

WangGLJoseph committed:
Change indentation for switch case to follow module coding standards

WangGLJoseph committed:
Add duke.txt with relative filePath

WangGLJoseph committed:
Update Duke with filePath to duke.txt

WangGLJoseph committed:
Handle creation of folder and file if not already present

WangGLJoseph committed:
Add functinality to write to duke.txt

WangGLJoseph committed:
Add skeletal method for copying from duke.txt and parsing it

WangGLJoseph committed:
Updated constructor of various task types

WangGLJoseph committed:
Revert constructor of various task types

WangGLJoseph committed:
Add parse method to convert duke.txt to taskArrayList

WangGLJoseph committed:
Fix bug where parse() was called twice incorrectly

WangGLJoseph committed:
Level-7: Add Save and Load duke.txt for Duke

WangGLJoseph committed:
Update Deadline and Event to parse task description as string of date

WangGLJoseph committed:
Level-8: Add Dates and Times functionality to Duke

WangGLJoseph committed:
Increment Duke display level

WangGLJoseph committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Increment Duke display level
  Level-8: Add Dates and Times functionality to Duke
  Update Deadline and Event to parse task description as string of date

# Conflicts:
#	src/main/java/Duke.java

WangGLJoseph committed:
Add skeletal classes for more OOP

WangGLJoseph committed:
Update skeletal class files to make Duke more OOP

kaushikkrdy commented on src/main/java/DukeParser.java:
```diff
@@ -0,0 +1,106 @@
+     * @param toParse the string to parse
+     * @return a taskArrayList made up of tasks
+     */
+    public static ArrayList<Task> parse(String toParse) {
```
Would it be possible to extract some of the code into another function to reduce code repetition and length of the parse method?

kaushikkrdy commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,186 @@
+     * @param input the string to be validated
+     * @return true if the string is a valid date
+     */
+    public boolean isValidDate(String input) {
```
Might I suggest using an inbuilt Java method like DateTimeFormatter or LocalDateTime to replace isValidDate() to improve code readability?

kaushikkrdy commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,186 @@
+        if (index > taskArrayList.size()) {
+            throw new DukeException("This task index is not in the task list!");
+        }
+        System.out.println(sandwich("Got it, I have deleted this task: "
```
Maybe you could refactor this code and similar methods to fit in line with SLAP Coding Quality? (Single Level of Abstraction Principle)

kaushikkrdy commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,152 @@
+        }
+    }
+
+    public void showLoadingError() {
```
Maybe a header comment would be useful here?

kaushikkrdy reviewed (commented):
LGTM! Code is well-written, maybe could refactor a little bit to be in line with coding quality standard in the textbook.

WangGLJoseph committed:
Put all classes into cs2103.duke package

WangGLJoseph committed:
Add JUnit Tests for Duke

WangGLJoseph committed:
Update duke.txt initialisation to enable use of .jar files

WangGLJoseph committed:
Add complete JavaDoc for all classes in Duke

WangGLJoseph committed:
Remove unnecessary use of the 'this' keyword

WangGLJoseph committed:
Level-9: Add find function for Duke

WangGLJoseph committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Remove unnecessary use of the 'this' keyword

WangGLJoseph committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Level-9: Add find function for Duke

# Conflicts:
#	src/main/java/cs2103/duke/Task.java
#	src/main/java/cs2103/duke/TaskList.java

WangGLJoseph committed:
Update README.md

Zzhiwei commented on src/main/java/cs2103/duke/Deadline.java:
```diff
@@ -7,12 +7,13 @@
  */
 public class Deadline extends Task {
-
+    protected int index;
```
I think it is better to use less general names and more specific names for better understanding by other readers.
For example, you can change the name here to indicate the context of the index.

Zzhiwei commented on src/main/java/cs2103/duke/Deadline.java:
```diff
@@ -7,12 +7,13 @@
 public class Deadline extends Task {
-
+    protected int index;
     protected String by;
```
Likewise for the variable "by", can change the name to be more specific.

Zzhiwei commented on src/main/java/cs2103/duke/Duke.java:
```diff
@@ -26,7 +24,7 @@ public Duke(String dukeFilePath) {
-    public void run() throws IOException {
+    public void run() {
         System.out.println(ui.showWelcome());
         boolean canExit = false;
```
Awesome! Good use of descriptive name for boolean variables!

Zzhiwei commented on src/main/java/cs2103/duke/Duke.java:
```diff
@@ -39,7 +37,6 @@ public void run() throws IOException {
-                    System.out.println("temp is: " + temp);
                     storage.overwriteFile(dukeFilePath, temp);
                     System.out.println(ui.showGoodbye());
                 } else { // check first input
```
Shouldn't comments be on separate lines as code?

Zzhiwei commented on src/main/java/cs2103/duke/DukeParser.java:
```diff
@@ -55,7 +55,7 @@ public DukeParser(String dukeFilePath) {
             switch (taskType) {
             case 'T':
```
Shouldn't case body contain less code? Maybe can try to put the code here into a function then call the function.

Zzhiwei commented on src/main/java/cs2103/duke/Task.java:
```diff
@@ -5,21 +5,27 @@
-     * @param description description of task.
+     * @param index       Index of the task, starting from 0.
+     * @param description Description of the task.
+
```
is this empty line necessary?

Zzhiwei commented on src/main/java/cs2103/duke/TaskList.java:
```diff
@@ -214,4 +215,32 @@ public String deleteTask(int index) throws DukeException {
+            throw new DukeException("Invalid keyword, please enter only one keyword without blanks");
+        } else {
+            ArrayList<Task> subList = new ArrayList<>();
+            for (Task task : taskArrayList) { // for each task in the taskArrayList
+                if (task.description.contains(keyword)) { // add task to subList if contains keyword
```
I think comments should be on separate line.

Zzhiwei reviewed (changes requested):
Great job! However, there are a few improvements that you may want to make.

WangGLJoseph committed:
Merge branch 'master' of https://github.com/WangGLJoseph/ip

* 'master' of https://github.com/WangGLJoseph/ip:
  Update README.md

WangGLJoseph committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

* add-gradle-support:
  Add Gradle support

WangGLJoseph committed:
Add DukeLauncher as part of JavaFX basics

CraveToCode commented on src/main/java/cs2103/duke/Event.java:
```diff
@@ -0,0 +1,25 @@
+ */
+public class Event extends Task {
+    protected int index;
+    protected String at;
```
Perhaps you could consider using a more descriptive variable name in place of `at`, so that other readers can more intuitively understand what `at` means. Maybe `eventDate` or `unparsedEventDate`?

CraveToCode commented on src/main/java/cs2103/duke/Storage.java:
```diff
@@ -0,0 +1,85 @@
+        File dukeFile = new File(dukeFilePath);
+        // creates directory if it does not exist
+        if (dukeFileDirectory.mkdir()) {
+            // dukeFile.mkdir();
```
Maybe this comment could be considered as redundant? The code line above it might be self-explanatory enough.

CraveToCode commented on src/main/java/cs2103/duke/Storage.java:
```diff
@@ -0,0 +1,85 @@
+        }
+        // creates file if it does not exist
+        if (dukeFile.createNewFile()) {
+            // dukeFile.createNewFile();
```
Maybe this comment could be considered as redundant? The code line above it might be self-explanatory enough.

CraveToCode commented on src/main/java/cs2103/duke/TaskList.java:
```diff
@@ -0,0 +1,246 @@
+                maxDay = 30;
+                break;
+            case 2:
+                if (isLeapYear) {
```
I like that you are accounting for leap years, nice!

CraveToCode commented on src/main/java/cs2103/duke/ToDo.java:
```diff
@@ -0,0 +1,20 @@
+ * This class encapsulates a Todo Task object, which inherits from the Task class.
+ */
+public class ToDo extends Task {
+    protected int index;
```
Perhaps you could consider renaming `index` such that it is more self-explanatory about what it is referring to.

CraveToCode commented on src/main/java/cs2103/duke/ToDo.java:
```diff
@@ -0,0 +1,20 @@
+    public ToDo(int index, String description) {
+        super(index, description);
+        this.index = index;
+        this.description = description;
```
Similar to the previous variable, perhaps you could consider renaming `description` such that it is more self-explanatory about what it is referring to.

CraveToCode commented on src/main/java/cs2103/duke/Ui.java:
```diff
@@ -0,0 +1,146 @@
+        case "done":  // first input is done, check second input for integer
+            if (scanner.hasNextInt()) {
+                int taskNum = scanner.nextInt();
+                return tasks.finishTask(taskNum);
+            } else throw new DukeException("unspecified task to mark as done");
+        case "delete":
+            if (scanner.hasNextInt()) {
+                int taskNum = scanner.nextInt();
+                return tasks.deleteTask(taskNum);
+            } else throw new DukeException("unspecified task to delete");
+        case "todo":
+            String todoName = scanner.nextLine();
+            if (todoName.trim().equals("")) {
+                throw new DukeException("No task description");
+            }
+            return tasks.addTask("todo", todoName, "");
+        case "deadline":
+            String[] deadlineTokens = scanner.nextLine().split("\\s*/by\\s*");
+            if (deadlineTokens.length == 0) {
+                throw new DukeException("No task description");
+            } else if (deadlineTokens.length == 1) {
+                throw new DukeException("No task deadline");
+            }
+            String deadlineName = deadlineTokens[0];
+            String deadlineReminder = deadlineTokens[1];
+            return tasks.addTask("deadline", deadlineName, deadlineReminder);
+        case "event":
+            String[] eventTokens = scanner.nextLine().split("\\s*/at\\s*");
+            if (eventTokens.length == 0) {
+                throw new DukeException("No task description");
+            } else if (eventTokens.length == 1) {
+                throw new DukeException("No task duration");
+            }
+            String eventName = eventTokens[0];
+            String eventReminder = eventTokens[1];
+            return tasks.addTask("event", eventName, eventReminder);
+        case "find":
+            if (scanner.hasNext()) {
+                return tasks.findTasks(scanner.nextLine());
+            } else throw new DukeException("unspecified keyword to search for");
```
Perhaps you could consider wrapping each case in a private method, for better code readability and easier code management later on.

CraveToCode reviewed (commented):
LGTM!
Just provided a few suggestions with regards to some nits on code quality.

WangGLJoseph committed:
Update README.md

WangGLJoseph committed:
Extract GUI logic out of Duke and into Main

WangGLJoseph committed:
Update README.md

WangGLJoseph committed:
Create new package for controllers

WangGLJoseph committed:
Merge branch 'master' of https://github.com/WangGLJoseph/ip

* 'master' of https://github.com/WangGLJoseph/ip:
  Update README.md
  Update README.md

WangGLJoseph committed:
Level-10: Add GUI for Duke

WangGLJoseph committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Level-10: Add GUI for Duke
  Create new package for controllers
  Extract GUI logic out of Duke and into Main
  Add DukeLauncher as part of JavaFX basics

WangGLJoseph committed:
Update Duke to use GUI

WangGLJoseph committed:
Set font in GUI as Courier New

WangGLJoseph committed:
Make Duke wait before closing the window

WangGLJoseph committed:
Add assertions for key assumptions in the code

WangGLJoseph committed:
Update method header comments to follow the coding standard

WangGLJoseph committed:
Update DialogBox to display more text

WangGLJoseph committed:
TaskList class: Apply SLAP to shorten methods of adding tasks

The addTask method implements the addition of
multiple types of tasks (i.e. ToDo, Deadline, Event).

This violates the Single Level of Abstraction principle and
makes the code difficult to understand or debug.

Implementing the add task methods separately allows us to call different
methods when adding a task of a particular type (e.g. addTodo).

Let's implement addTodo, addDeadline, and addEvent as separate methods.

This reduces the risk of calling the wrong add task method.

WangGLJoseph committed:
Update .gitignore to ignore data/duke.txt

WangGLJoseph committed:
Remove data within data/duke.txt

WangGLJoseph committed:
DukeParser class: Apply SLAP to shorten methods of parsing data

The parse method implements the parsing of all data from duke.txt into
an ArrayList. The method handles the parsing of all types of tasks.

This violates the Single Level of Abstraction principle and
makes the code difficult to understand or debug.

Implementing each parse task methods separately allows us to
call different methods when parsing one type of task (e.g. addTodo).

Let's implement parseTodo, parseDeadline, and parseEvent
as separate methods.

This reduces the risk of calling the wrong parse task method.

WangGLJoseph committed:
Merge pull request #2 from WangGLJoseph/branch-A-Assertions

Add assertions for key assumptions in the code

WangGLJoseph committed:
Merge branch 'master' of https://github.com/WangGLJoseph/ip

* 'master' of https://github.com/WangGLJoseph/ip:
  Add assertions for key assumptions in the code

WangGLJoseph committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add assertions for key assumptions in the code

WangGLJoseph committed:
Merge pull request #3 from WangGLJoseph/branch-A-CodeQuality

Apply SLAP to TaskList and DukeParser

WangGLJoseph committed:
Merge branch 'master' of https://github.com/WangGLJoseph/ip

* 'master' of https://github.com/WangGLJoseph/ip:
  DukeParser class: Apply SLAP to shorten methods of parsing data
  Remove data within data/duke.txt
  Update .gitignore to ignore data/duke.txt
  TaskList class: Apply SLAP to shorten methods of adding tasks
  Update DialogBox to display more text
  Update method header comments to follow the coding standard

WangGLJoseph committed:
Add functionality to undo the last undo-able command

WangGLJoseph committed:
Add Ui.png

WangGLJoseph committed:
Update user guide for Duke

WangGLJoseph committed:
Set theme jekyll-theme-time-machine

WangGLJoseph committed:
Update formatting in user guide

WangGLJoseph committed:
Merge branch 'master' of https://github.com/WangGLJoseph/ip

* 'master' of https://github.com/WangGLJoseph/ip:
  Set theme jekyll-theme-time-machine

WangGLJoseph committed:
Remove emojis

WangGLJoseph committed:
Update goodbye message
